Adding Sanity Checks

michaelavs opened the pull request:
Adding sanity checks from ncomp to f2py

michaelavs commented:
@erogluorhan @pilotchute @anissa111 
Hey everyone,
This is just a quick start at adding the sanity checks to f2py. I started with linint2 which has a couple of functions in the wrapper that I couldn't locate in ncomp (_linint1, linint1, linint2pts). Were they renamed from ncomp to f2py? Additionally, would you all like me to also update naming conventions to use lat/lon with this PR? Let me know what you think so far about what looks good and what needs changing!

michaelavs commented:
@pilotchute I added a very basic first draft of a sanity check library, let me know what you think about the format. I haven't been able to fully test them yet, but I want to make sure this is the direction we want to go in first before getting too deep.

pilotchute commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -27,6 +27,110 @@ def _linint1(xi, fi, xo, icycx, msg_py, shape):
+        rightmost two dimensions which will have the same dimension
+        sizes as the lengths of yo and xo. The return type will be
+        double if fi is double, and float otherwise.
+    '''
```
Help text is not needed in the inner wrapper. This appears to just duplicate the text in the outer wrapper.

pilotchute commented on src/geocat/f2py/sanity_checks.py:
```diff
@@ -0,0 +1,74 @@
+        if (yi is None) | (xi is None):
+            raise CoordinateError(
+                "Arguments yi or xi must be provided explicitly unless fi is an xarray.DataArray.")
+
```
This is not quite what we want, let's set up a time to work on it together. 
We don't want our sanity_check library to simply pull function specific checks out of the functions, we want to make a general use sanity_checker that can be used by many functions for many datatypes and input variables.

pilotchute reviewed (changes requested):
Let's work together on this, there's a design pattern I want to help you learn.

michaelavs commented:
@erogluorhan @pilotchute @anissa111 
I wanted to go over the changes I made to the check functions with everyone before merging because I want to make sure we're all on the same page with them and that the changes are useful overall.

**1. Updating the ```is_xarray``` check**
When testing this check, I found that the original version wasn't necessarily doing a good job at what it was meant to catch. The primary issue I saw was that it would raise and exception even if the data was the correct format, which obviously causes the code to stop running where it is. To prevent this, I added an 'else' statement to ```if is_xarray is True``` and further expanded ```if is_xarray is False``` to instead print 'is not xarray, reformatting' and incorporated this check into the rcm2points_wrapper. The wrapper will now take the return of 'True' and reformat the data array into an xarray.DataArray using the method used found in lines 110-120. I Have not cross checked if this method occurs in every wrapper function, but if it does, then this check can easily be integrated with the existing functions. If it doesn't, I had played around with adding the reformatting function into the check and just keeping it all in one place. I did ultimately decide against this in case there were sizable differences across each wrapper function, but I would be willing to look into it more if that's the path we want to go down.

**2. Updating ```data_type``` check**
As with the previous point, while testing this check, I found it wasn't very intuitive in how it was checking. When using ```dtype``` previously, the function would look at the bytes of the data instead of the actual type. Because this check was meant to make sure the user input an np.ndarray (as outlined in the wrapper), it seemed like it would be more worth while to be checking the actual type over dtype. However, if it is preferred by everyone else to check against dtype instead, I can change this back and modify the code to make sure this is communicated and implemented effectively. 

**3. Adding brief descriptions of most of the checks**
I added in brief descriptions to the majority of the checks. The only ones I could not figure out how to test were ```comparison``` and ```is_none``` because I wasn't too sure of where/how to implement them in rcm2points. Overall, I figured that these checks will not be interacted with by the user as much as they will be interacted with by the team, so keeping the descriptions short and to the point would be fine in this case. If you guys would like to add anything to the descriptions, or have any input for comparison and is_none, I am definitely open to revising/adding to them!
  
I think that covers all of the major updates I made in this code, but if you have any questions/comments/concerns please let me know! The checks have also been added to rcm2points and have been running successfully via the test script, so you can see how/where they may be used in other wrapper functions.

erogluorhan reviewed (changes requested):
Even though I agree with implementing common data checks in such a module for repetitive use, I have a number of concerns with the current implementation here as follows:

1. As can be seen from my inline comments as well, the flow of the functions this check is currently applied to (linint2_wrapper, rcm2points) seem to have involved into a less readable/intuitive way.
2. Some data flexibilities seem to have been diminished (`xo` could be xarray and numpy before but strictly numpy now), which results in failing test cases now. I also pointed this out in inline comments.
3. See [GeoCAT-comp Issue 126](https://github.com/NCAR/geocat-comp/issues/126) as a reference that we need to work on the boiler plate of our functions. I believe we should handle this checks.py, its application to our codes, and our boiler plates all together.
4. Whenever we finalize such a study, I think we should better apply that to all the existing functions within the PR to avoid need for a further revisit.

michaelavs commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -98,23 +98,33 @@ def linint1(fi, xo, xi=None, icycx=0, msg_py=None):
+        check(yi, data_type=np.ndarray)
+
+    # Check if fi is xarray.DataArray, if not change it
+    if check(fi, is_xarray=False) == True:
```
I agree. When writing this line, I wanted to make sure that by calling ```check(fi, is_xarray=False)``` there wouldn't be an issue if user used```check(fi, is_xarray=True)```. However, something I wanted to double check is the necessity of having ```is_xarray``` as a True/False statement. It would seem that we would only need it to be defined for one option (only True or only False) and then have the wrapper function defined for that (e.g. only using ```check(fi, is_xarray=True)``` or ```check(fi, is_xarray=False)``` in the wrapper). That being said, I think this may also need to be reformatted in order to resolve your second comment above, because it is true that the wrapper functions can handle both the xarray.DataArray and numpy.ndarray. I believe that Austin and I used ```is_xarray``` due to the original ncomp wrapper check saying that the variable type should be in is xarray.DataArray only. 

With respect to your first, third, and fourth point above, and considering the other changes requested for this PR, I think we should re-visit this PR as a team to make sure it is going in the right direction for this repository. In the mean time, I will work on the changes outlined below, and ensuring they work with the two wrappers I have been testing implementation in.

erogluorhan replied:
> With respect to your first, third, and fourth point above, and considering the other changes requested for this PR, I think we should re-visit this PR as a team to make sure it is going in the right direction for this repository. In the mean time, I will work on the changes outlined below, and ensuring they work with the two wrappers I have been testing implementation in.

I agree. Let us discuss these in our weekly meeting, and in the meantime, I will make a document to write down major points that would need discussion/modification etc.

michaelavs commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -98,23 +98,33 @@ def linint1(fi, xo, xi=None, icycx=0, msg_py=None):
+    # Check if fi is xarray.DataArray, if not change it
+    if check(fi, is_xarray=False) == True:
+        if not isinstance(fi, xr.DataArray):
+            if (xi is None):
```
Would it be best to remove the lines for checking xi starting at 110, or is there a reason we would want to check xi when checking fi? Comparing with the checks starting on 105, I think keeping line 105 checks and removing the line 110 checks should yield the same results as the original check in 112-115.

erogluorhan commented:
@michaelavs The current code pushed in this branch has traces of merge conflict strings and so pops up syntax errors (Therefore cannot pass tests, too). If you are still working on it, that's okay, but if you intended it to be for my next review, please fix them and then I am happy to give another review cycle.

michaelavs commented:
> @michaelavs The current code pushed in this branch has traces of merge conflict strings and so pops up syntax errors (Therefore cannot pass tests, too). If you are still working on it, that's okay, but if you intended it to be for my next review, please fix them and then I am happy to give another review cycle.

Should be fixed now!

erogluorhan commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -98,23 +98,33 @@ def linint1(fi, xo, xi=None, icycx=0, msg_py=None):
-        ).chunk(fi_chunk)
+
+    # Check input data type
+    check(xo, data_type=np.ndarray)
```
Please remove the checks for `xo` and `yo` as they can be either numpy or xarray.

erogluorhan commented on src/geocat/f2py/checks.py:
```diff
@@ -0,0 +1,164 @@
+            raise Exception(var_name + " failed shape check")
+
+    if is_none is not None:
+        if is_none is True:
```
Since `is_none` is already a boolean, there is no need for additional `is True`. This is also applicable to other blocks, if they written similarly. The whole block here could then be:

```
   if is_none is not None:
        if is_none:
            if user_variable is not None:
                raise Exception(var_name + " is not None")
        else:
            if user_variable is None:
                raise Exception(var_name + " is None")
```

erogluorhan commented on src/geocat/f2py/checks.py:
```diff
@@ -0,0 +1,164 @@
+
+xt = xi
+
+def dtype_2d(xi=None, yi=None):
```
These two functions `dtype_2d ` and `dtype_1d ` are very problem-specific and do not look like commony-applicable. They may be given up for the time until we see multiple common cases that would need such checking functions.

erogluorhan commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -296,61 +282,61 @@ def linint2(fi, xo, yo, xi=None, yi=None, icycx=0, msg_py=None):
-                "linint2: Arguments xi and yi must be provided explicitly unless fi is an xarray.DataArray."
-            )
+    # Check data type for input variables
+    check(xo, data_type=np.ndarray) # ToDo: check necessity
```
Please remove these checks for `xo` and `yo` as they can be either xarray or numpy without an issue.

erogluorhan commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -296,61 +282,61 @@ def linint2(fi, xo, yo, xi=None, yi=None, icycx=0, msg_py=None):
+    check(xo, data_type=np.ndarray) # ToDo: check necessity
+    check(yo, data_type=np.ndarray)
+
+    if (xi is not None) | (yi is not None):
```
If `fi` is `Xarray` already (which is checked for in Line 320), we are not really concerned with `xi` and `yi`. Our concern for `xi` and `yi` to be `np.ndarray` starts when `fi` is not xarray. Therefore, these check for `xi` and `yi` are redundant here. It is enough to check them only after we see `fi` is not xarray, i.e. it is numpy. This is possible starting Line 322.

erogluorhan commented on src/geocat/f2py/linint2_wrapper.py:
```diff
@@ -296,61 +282,61 @@ def linint2(fi, xo, yo, xi=None, yi=None, icycx=0, msg_py=None):
+        check(xi, data_type=np.ndarray)
+        check(yi, data_type=np.ndarray)
+
+    if check(fi, is_xarray=False) == True:
```
As I mentioned before, there are two same checks here that checks whether `fi` is an Xarray. The correct form should be:

```
if not isinstance(fi, xr.DataArray):
    if (xi is None) | (yi is None):
         raise CoordinateError( ...
    ...
```
or 

```
if check(fi, is_xarray=False):
    if (xi is None) | (yi is None):
         raise CoordinateError( ...
    ...
```

erogluorhan reviewed (changes requested):
I have made multiple inline comments. Please try to address them.

When they are addressed, almost none of these check features will be existing in, e.g. `linint2_wrapper`, but that's okay, we are implementing common checking features here that may or maynot be applicable for specific functions. In other words, if the function boiler plate is good enough, and if there is not many condition-checks in that function, which would be addressed with this `check` module, that's okay.

pilotchute reviewed (approved):
Approved per team coding on hackday
